### PR TITLE
[BugFix] fix and refactor logical of inference bucket number

### PR DESCRIPTION
**Pull request by murphyatwork**
Fixes #issue

1. Unify the bucket number inference code to `Table::inferDistribution`
2. Not change default bucket number, so `show create table` would not display the bucket number  if not specified manually 

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 1]
```diff
@@ -1259,6 +1259,29 @@ public void setPlanContext(MVRewriteContextCache mvRewriteContextCache) {
...
+     */
+    @Override
+    public void inferDistribution(DistributionInfo info) {
+        if (info.getBucketNum() == 0) {
+            int inferredBucketNum = 0;
+            for (BaseTableInfo base : getBaseTableInfos()) {
+                if (base.getTable().isNativeTableOrMaterializedView()) {
+                    OlapTable olapTable = (OlapTable) base.getTable();
+                    DistributionInfo dist = olapTable.getDefaultDistributionInfo();
+                    inferredBucketNum = Math.max(inferredBucketNum, dist.getBucketNum());
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1275]: The base table bucket number may be too larger for MV?
If the data can aggregated to 10x times, the bucket number in the base table is not suitable for the MV

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1275]: i know, but it's better than now

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|5|100.00%|[]|

**Review by ABingHuang — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|5|100.00%|[]|

**Review by packy92 — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**